File: 337_House_Robber_III.go
// 2020/6/13
package main

import (
	"fmt"
	"time"
)

func main() {
	start := time.Now()
	head := &TreeNode{}
	//fmt.Println(rob(head) == 0)
	//head = &TreeNode{3, nil, nil}
	//fmt.Println(rob(head) == 3)
	head = &TreeNode{3, &TreeNode{2, nil, &TreeNode{3, nil, nil}}, &TreeNode{3, nil, &TreeNode{1, nil, nil}}}
	fmt.Println(rob(head) == 7)
	elapsed := time.Now().Sub(start)
	fmt.Println(elapsed)
}

//Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

var memo map[*TreeNode]int

func robMemo(node *TreeNode) int {
	if node == nil {
		return 0
	}
	if v, ok := memo[node]; ok {
		return v
	}

	grandChildren := 0
	if node.Left != nil {
		grandChildren += robMemo(node.Left.Left) + robMemo(node.Left.Right)
	}
	if node.Right != nil {
		grandChildren += robMemo(node.Right.Left) + robMemo(node.Right.Right)
	}

	memo[node] = node.Val + grandChildren
	child := robMemo(node.Left) + robMemo(node.Right)
	if child > memo[node] {
		memo[node] = child
	}
	return memo[node]
}

func rob(root *TreeNode) int {
	memo = make(map[*TreeNode]int)
	return robMemo(root)
}

File: 337_House_Robber_III.py
# 2020/6/13
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution(object):
    def rob(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        def robRecursive(node):
            if not node:
                return 0

            grandchildren = (
                robRecursive(node.left.left) +
                robRecursive(node.left.right) if node.left else 0
            ) + (
                robRecursive(node.right.left) +
                robRecursive(node.right.right) if node.right else 0
            )

            return max(
                node.val + grandchildren,
                robRecursive(node.left) + robRecursive(node.right))

        def robMemoization(node):
            if not node:
                return 0
            elif node in memoize:
                return memoize[node]

            grandchildren = (
                robMemoization(node.left.left) +
                robMemoization(node.left.right) if node.left else 0
            ) + (
                robMemoization(node.right.left) +
                robMemoization(node.right.right) if node.right else 0
            )

            memoize[node] = max(
                node.val + grandchildren,
                robMemoization(node.left) + robMemoization(node.right))
            return memoize[node]

        memoize = dict()
        return robMemoization(root)
